### PR TITLE
remove role/rolebinding finalizer when manifestworks are cleaned

### PR DESCRIPTION
**Pull request by qiujian16**
remove finalizer on rbac when all manifestworks are cleaned (samiliar from pr https://github.com/open-cluster-management/work/pull/23) but i think it is better to move to registration repo.

### CODE REVIEW

File: pkg/hub/rbacfinalizerdeletion/controller.go [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+			return err
+		}
+
+		if len(works) != 0 {
+			return fmt.Errorf("Still having %d works in the cluster namespace %s", len(works), ns.Name)
+		}
+	}
+
+	// remove finalizer from role/rolebinding
+	if terminated(role) {
```
Comment (deads2k) [inline pkg/hub/rbacfinalizerdeletion/controller.go]: nit: I prefer `pendingFinalization`

File: pkg/hub/rbacfinalizerdeletion/controller.go [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+	ns, err := m.namespaceLister.Get(namespace)
+	if err != nil {
+		return err
+	}
+
+	// cluster is deleted or being deleted
+	role, rolebinding, err := m.getRoleAndRoleBinding(namespace, name)
```
Comment (deads2k) [inline pkg/hub/rbacfinalizerdeletion/controller.go:90]: followup: I think you should iterate over every role and rolebinding in the namespace.  Keying to one name is problematic long term

Comment (deads2k) [inline pkg/hub/rbacfinalizerdeletion/controller.go:90]: Also note, if you had a `CriticalResource` you wouldn't be writing this code at all and we would be inverting the checks.

Comment (deads2k) [inline pkg/hub/rbacfinalizerdeletion/controller.go:90]: another followup: if you first traverse the roles, then traverse the rolebindings, I think the code becomes more readable.  Special cases defy user-expectations.

**Comment by deads2k** [comment]
/lgtm
/approve

/hold

holding in case @pmorie has comments.

File: pkg/hub/rbacfinalizerdeletion/controller.go [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+	role *rbacv1.Role, rolebinding *rbacv1.RoleBinding, ns *corev1.Namespace, cluster *clusterv1.ManagedCluster) error {
+	// Skip if neither role nor rolebinding has the finalizer
+	if !hasFinalizer(role, manifestWorkFinalizer) && !hasFinalizer(rolebinding, manifestWorkFinalizer) {
+		return nil
+	}
+
+	// There are twp possible cases that we need to remove finalizers on role/rolebindings based on
+	// clean of manifestworks.
+	// 1. The namespace is finalizing.
+	// 2. The cluster is finalizing but namepsace fails to be deleted.
```
Comment (pmorie) [inline pkg/hub/rbacfinalizerdeletion/controller.go]: typo: namespace

File: test/integration/managedcluster_deletiong_test.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+	clusterv1 "github.com/open-cluster-management/api/cluster/v1"
+
+	testinghelpers "github.com/open-cluster-management/registration/pkg/helpers/testing"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/rand"
+)
+
+var _ = ginkgo.Describe("Cluster deletion", func() {
```
Comment (pmorie) [inline test/integration/managedcluster_deletiong_test.go]: note: typo in name of this file: `deletiong`

**Comment by pmorie** [comment]
Found a couple nits, you can fix them in a follow-up.

/hold cancel

**Pull request merged**